src/mfs.jl

```diff
@@ -0,0 +1,170 @@
+    best_dist: norm of the best shock found so far
+    dim: dimension of the system
+    total_iterations: number of random pertubations to try, default = 10000
+
+Output:
+    best_shock: vector of the best shock found
+    best_dist: norm of the best shock found
+
+
+"""
+
+
+function mfs_brute_force(mapper::AttractorMapper, X, best_shock, best_dist, dim, id_x,  total_iterations=10000)
+    
+    radius_dif = 100.
+
+    for _ in 1:total_iterations
+
+        pertubation = random_point_on_sphere((best_dist-best_dist/radius_dif), dim)
+        new_shock = X + pertubation
+        
+        if !(id_x == mapper(new_shock))
+            best_dist = LinearAlgebra.norm(pertubation)
+            best_shock = pertubation
+        end
+    end
+
+    return best_shock, best_dist
+end
+
+
+
+
+function mfs_objective(perturbation, X, idX, mapper::AttractorMapper, penalty=1000.0) 
+
+    dist = LinearAlgebra.norm(perturbation)
+    if dist == 0
+        return penalty
+    end
+    if mapper(X + perturbation) == idX
+        return dist + penalty
+    else
+        return dist
+    end
+end
+
+
+"""
+    minimal_fatal_shock(mapper::AttractorMapper, X, search_area; dimension, total_iterations = 10000, algorithm = :random)  
+
+Runs the minimal fatal shock algorithm on the initial point X.
+Three choises available for the algorithm (keywords algorithm = ...): random, bboxopt and combined.
+In "random" algorithm it generates random pertubation of X and then trying to find a better one by using the brute force method.
+In "bboxopt" it uses algorithms decribed in BlackBoxOpt.jl julia package to perform derivative free constraint optimization iof finding minimal fatal shock.
+The "combined" method is used to initially run "random" algorithm and then provide obtained guess to the "bboxopt" method for better accuracy and optimization. 
+
+
+
+Input:
+    mapper: AttractorMapper
+    search_area: array of tuples, by specifying single value e.g. [(-1.5, 1.5)] you specify this value for each of the variable, or you may specify separetely 
+search area for each of your states by, for example in 2 dimensions, [(-1.5, 1.5), (-1, 1)]
+    X: initial point to test
+    dimension: dimension of your input system
+    total_iterations: number of iterations to run each step of the algorithm, default = 10000
+    algorithm: random, bboxopt or combined.  
+Output:
+    best_shock: vector of the best shock found
+    best_dist: norm of the best shock found
+
+
+"""
+
+function minimal_fatal_shock(mapper::AttractorMapper, X, search_area; dimension, total_iterations = 10000, algorithm = :random)
+
+
+    id_x = mapper(X)
+    
+    if length(search_area) == 1
+        search_area = [search_area[1] for _ in 1:dimension]
+    end
+
+    function objective_function(perturbation)  # we need this declaration here due to BlackBoxOpt specifics of input
+        return mfs_objective(perturbation, X, id_x, mapper)
+    end
+
+    if algorithm == :random 
```

Comment:
I said that we should use multiple dispatch. Create types that correspond to the algorithm possiblities and in these types the user can provide deciding arguments. Like, which algorithm from blackboxoptim to use, or how large the penantly wshould be, or how many iterations to do. Everything should be tunable by the user by tuning the fields of these new types that decide the algorithm.